eval/compile_effect.go

```diff
@@ -358,23 +328,35 @@ func allTrue(vs []interface{}) bool {
 	return true
 }
 
-func (cp *compiler) assignment(n *parse.Assignment) effectOpBody {
+func (cp *compiler) assignment(n *parse.Assignment, temporary bool) effectOpBody {
 	variablesOp, restOp := cp.lvaluesOp(n.Left)
 	valuesOp := cp.compoundOp(n.Right)
-	return &assignmentOp{variablesOp, restOp, valuesOp}
+	return &assignmentOp{variablesOp, restOp, valuesOp, temporary, false, nil, nil}
 }
 
 // ErrMoreThanOneRest is returned when the LHS of an assignment contains more
 // than one rest variables.
 var ErrMoreThanOneRest = errors.New("more than one @ lvalue")
 
+// assignmentOp assign values to variables. If the assignment is
+// temporary, calling it again will restore original values.
 type assignmentOp struct {
 	variablesOp lvaluesOp
 	restOp      lvaluesOp
 	valuesOp    valuesOp
+	temporary   bool
+	assigned    bool
+	origVars    []vars.Var
+	origVals    []interface{}
 }
 
 func (op *assignmentOp) invoke(fm *Frame) (errRet error) {
+	if op.temporary && op.assigned {
```

Comment:
This feels like fitting a square peg into a round whole: since `assignmentOp` can now do both assignment and restoration, `effectOpBody` is not an expressive enough interface for it. For instance, you can define a `revertableEffectOpBody` which can look like:

```go
type revertableEffectOpBody interface {
    invoke(fm *Frame) (revert effectOp, err error)
}
```

Or maybe you want `assignmentOp` to continue satisfying the `effectOpBody` interface (which is useful when doing permanent assignments as you don't need to revert), and let it satisfy another interface in addition; in that case you would give the method another name:

```go
type revertableEffectOpBody interface {
    invokeRevertable(fm *Frame) (revert effectOp, err error)
}
```

I'm not sure which option will work the best; you might want to play with those options and maybe be creative.